File: 198.house-robber.py
#
# @lc app=leetcode id=198 lang=python3
#
# [198] House Robber
#

# @lc code=start
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        cache = [-1] * len(nums)
        def backtracking(index):
            if index >= len(nums):
                return 0
            if cache[index] != -1:
                return cache[index]
            cache[index] = max(nums[index] + backtracking(index + 2), backtracking(index + 1))
            return cache[index]
        return backtracking(0)
        
# @lc code=end

File: 198.house-robber.java
/*
 * @lc app=leetcode id=198 lang=java
 *
 * [198] House Robber
 */

// @lc code=start

class Solution {

    public int rob(int[] nums) {
        int rob1 = 0, rob2 = 0;
        for (int num : nums) {
            int temp = Math.max(rob1 + num, rob2);
            rob1 = rob2;
            rob2 = temp;
        }
        return rob2;
    }
}
// @lc code=end

